terraform/state.go

```diff
@@ -185,6 +185,41 @@ func (s *State) moduleByPath(path []string) *ModuleState {
 	return nil
 }
 
+// GetResourceStateFromResourceAddress accept a ResourceAddress and
+// ResourceType and return the ResourceState that correspoding to the
+// ResourceAddress
+func (s *State) GetResourceStateFromResourceAddress(rsa *ResourceAddress, resourceType string) (*ResourceState, error) {
+
+	// Getting ModuleState from resource address path
+	mod := s.ModuleByPath(rsa.Path)
+	module := strings.Join(rsa.Path, ".")
+	if mod == nil {
+		return nil, fmt.Errorf(
+			"The module %s could not be found. There is nothing to taint.", module)
+	}
+
+	var name string
+	if rsa.Index != -1 {
+		name = strings.Join([]string{resourceType, rsa.Name, strconv.Itoa(rsa.Index)}, ".")
+	} else {
+		name = strings.Join([]string{resourceType, rsa.Name}, ".")
+	}
+
+	// If there are no resources in this module, it is an error
+	if len(mod.Resources) == 0 {
+		return nil, fmt.Errorf(
+			"The module %s has no resources. There is nothing to taint.", module)
```

Comment:
I assume this error message was from before you did some refactoring... with the current approach, it's not guaranteed that the caller here is retrieving the resource state in order to taint it.

Given that this function is general and in core, I'd suggest that rather than returning an error we instead return `nil` in all of these error cases and then produce suitable error messages in the UI code, where we have more context available about what operation is being performed.